pageserver/src/tenant.rs

```diff
@@ -3003,6 +3029,32 @@ impl Tenant {
         Ok(tl)
     }
 
+    /// Helper for unit tests to branch a timeline with some pre-loaded states.
+    #[cfg(test)]
+    #[allow(clippy::too_many_arguments)]
+    pub async fn branch_timeline_test_with_states(
+        &self,
+        src_timeline: &Arc<Timeline>,
+        dst_id: TimelineId,
+        start_lsn: Option<Lsn>,
+        ctx: &RequestContext,
+        delta_layer_desc: Vec<Vec<(pageserver_api::key::Key, Lsn, crate::repository::Value)>>,
+        image_layer_desc: Vec<(Lsn, Vec<(pageserver_api::key::Key, bytes::Bytes)>)>,
+        end_lsn: Lsn,
+    ) -> anyhow::Result<Arc<Timeline>> {
+        let tline = self
+            .branch_timeline_test(src_timeline, dst_id, start_lsn, ctx)
+            .await?;
+        tline.force_advance_lsn(end_lsn);
```

Comment:
I wonder if there is ever a situation where we would want this to be less than what was written to the layer files? At least an assertion for this would be easy, or direct calculation of the `end_lsn` from the layer descriptions.

Reply:
The `force_create_*` functions will check key LSNs against last_record_lsn, so it's already checked. A direct calculation might be too implicit because we need an LSN for read operations (which is usually the same as last_record_lsn) and it's better developers specify them beforehand.